src/Mod/Measure/Gui/ViewProviderMeasureBase.cpp

```diff
@@ -0,0 +1,577 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include "Gui/Application.h"
+#include "Gui/MDIView.h"
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+# include <boost_signals2.hpp>
+# include <boost/signals2/connection.hpp>
+
+# include <Inventor/actions/SoGetMatrixAction.h>
+# include <Inventor/nodes/SoAnnotation.h>
+# include <Inventor/nodes/SoBaseColor.h>
+# include <Inventor/nodes/SoCoordinate3.h>
+# include <Inventor/nodes/SoCamera.h>
+# include <Inventor/nodes/SoDrawStyle.h>
+# include <Inventor/nodes/SoIndexedLineSet.h>
+# include <Inventor/nodes/SoMarkerSet.h>
+# include <Inventor/nodes/SoPickStyle.h>
+# include <Inventor/draggers/SoTranslate2Dragger.h>
+# include <Inventor/engines/SoComposeMatrix.h>
+# include <Inventor/engines/SoTransformVec3f.h>
+# include <Inventor/engines/SoConcatenate.h>
+#endif
+
+#include <App/DocumentObject.h>
+#include <Base/Console.h>
+#include <Gui/Document.h>
+#include <Gui/ViewParams.h>
+#include <Gui/Inventor/MarkerBitmaps.h>
+#include <Gui/View3DInventor.h>
+#include <Gui/View3DInventorViewer.h>
+
+#include <Mod/Measure/App/Preferences.h>
+#include "ViewProviderMeasureBase.h"
+
+
+using namespace MeasureGui;
+using namespace Measure;
+namespace bp = boost::placeholders;
+
+//NOLINTBEGIN
+PROPERTY_SOURCE(MeasureGui::ViewProviderMeasureBase, Gui::ViewProviderDocumentObject)
+//NOLINTEND
+
+ViewProviderMeasureBase::ViewProviderMeasureBase()
+{
+    static const char *agroup = "Appearance";
+//NOLINTBEGIN
+    ADD_PROPERTY_TYPE(TextColor, (Preferences::defaultTextColor()), agroup, App::Prop_None, "Color for the measurement text");
+    ADD_PROPERTY_TYPE(TextBackgroundColor, (Preferences::defaultTextBackgroundColor()), agroup, App::Prop_None, "Color for the measurement text background");
+    ADD_PROPERTY_TYPE(LineColor, (Preferences::defaultLineColor()), agroup, App::Prop_None, "Color for the measurement lines");
+    ADD_PROPERTY_TYPE(FontSize, (Preferences::defaultFontSize()), agroup, App::Prop_None, "Size of measurement text");
+//NOLINTEND
+
+    // setupAnnoSceneGraph() - sets up the annotation scene graph
+    pLabel = new Gui::SoFrameLabel();
+    pLabel->ref();
+    pColor = new SoBaseColor();
+    pColor->ref();
+    pLabelTranslation = new SoTransform();
+    pLabelTranslation->ref();
+
+    auto ps = getSoPickStyle();
+
+    // Dragger
+    SoSeparator* dragSeparator = new SoSeparator();
+    pDragger = new SoTranslate2Dragger();
+    pDragger->ref();
+    pDraggerOrientation = new SoTransform();
+    pDraggerOrientation->ref();
+    dragSeparator->addChild(pDraggerOrientation);
+    dragSeparator->addChild(pDragger);
+
+    // Transform drag location by dragger local orientation and connect to labelTranslation
+    auto matrixEngine = new SoComposeMatrix();
+    matrixEngine->rotation.connectFrom(&pDraggerOrientation->rotation);
+    auto transformEngine = new SoTransformVec3f();

```

Comment:
Memory leak

Reply:
Not quite sure here, shouldn't openinventor's reference counting take care of this? The [documentation](http://www-evasion.imag.fr/Membres/Francois.Faure/doc/inventorMentor/sgi_html/ch13.html#id5493141) says that the reference count of an engine is increased if an output is connected to another field. Therefor the lifetime of matrixEngine should follow pLabelTranslation, am i missing something?

Reply:
I'm not sure either but you are probably right. I never worked a lot with engines and thus don't know its details.